### PR TITLE
[Time Conductor] Add time conductor

### PR DESCRIPTION
**Pull request by VWoeltjen**
Adds a bundle which adds a time conductor, WTD-1515. Summary of approach used:

* Define three additional (optional) properties of a `TelemetryRequest`:
  * `start`: First time of interest
  * `end`: Last time of interest
  * `size`: Total number of data points requested (implied preference is near the end), for latest-data style requests (e.g. `size=1`)
* Add a controller (in the Angular sense) to go along with the template that @charlesh88 prepared previously, making it a usable control (in the put-data-in get-data-out sense)
* Add a representer which inserts the control mentioned above at the bottom of whatever the outermost view is that gets represented
* Add a class to maintain current time conductor state, shared between the representer (above) and telemetry decorator (below) by way of a service (so, effectively a singleton)
* Decorate telemetry capabilities such that requests and subscription both have start and end times added, based on the state of the time conductor
  * Currently, there's a check to see if a size has been specified, to distinguish plot-like requesters from fixed-position-like requesters. The former can handle queries larger than the displayable region (which is nice for time conductor driven panning) while the latter really just wants the single last displayable data point.
* Added event listeners to Plot and Fixed Position view to listen for changes to time conductor state and requery when these occur

Opening this pull request a bit early to solicit feedback at the lines-of-code level. Still to-do:

* Unit tests for `TimeConductorController` 
* Punish myself for creating a class called `TimeConductorController`
* Add domain selector
  * Plan to do this by monitoring `telemetry` capabilities as they are instantiated to get their metadata
* Remove the weird check to see if `size` has been specified, and always just query for the displayable region. This means plot will need to requery and we'll lose the smooth scroll effect, but it avoids introducing subtle weird rules that views would need to account for.
* Scrolling list view
  * This needs to be updated anyway to reconcile real-time and historical telemetry (which also needs to happen for this view to synchronize usefully with the time conductor)
* User feedback (e.g. red borders) when typed dates for start/end bounds of time conductor cannot be parsed

### CODE REVIEW

File: platform/features/conductor/src/ConductorCapabilityDecorator.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+         * capabilities have their requests mediated by the time conductor.
+         *
+         * @constructor
+         * @memberof platform/features/conductor
+         * @implements {CapabilityService}
+         * @param {platform/features/conductor.ConductorService} conductorServe
+         *        the service which exposes the global time conductor
+         * @param {CapabilityService} capabilityService the decorated service
+         */
+        function ConductorCapabilityDecorator(conductorService, capabilityService) {
```
Comment (larkin) [inline platform/features/conductor/src/ConductorCapabilityDecorator.js]: Note: I haven't finished reading the PR, but I'm surprised that we have to decorate both the capability service and the telemetry service-- why can't the telemetry service decorator do everything?

File: platform/features/conductor/src/ConductorRepresenter.js [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+         *        service which provides the active time conductor
+         * @param $compile Angular's $compile
+         * @param {ViewDefinition[]} views all defined views
+         * @param {Scope} the scope of the representation
+         * @param element the jqLite-wrapped representation element
+         */
+        function ConductorRepresenter(conductorService, $compile, views, scope, element) {
+            var conductorScope;
+
+            // Angular doesn't like objects to retain references to scopes
```
Comment (larkin) [inline platform/features/conductor/src/ConductorRepresenter.js]: As this is a closure an instance of the ConductorRepresenter will still retain a reference to the scope-- might as well be consistent and store as `this.scope`.

File: platform/features/conductor/src/ConductorCapabilityDecorator.js (hunk 1 above)
Comment (VWoeltjen) [inline platform/features/conductor/src/ConductorCapabilityDecorator.js]: So, it's the telemetry _capability_ that gets decorated, by way of the `capabilityService`

That said, it probably would be simpler to decorate the `telemetryService` directly and skip capabilities altogether. Byproduct of doing it this way would be that:

1. Domain objects which explicitly override their `telemetry` capability to not use the `telemetryService` will not use the time conductor; we don't have anything like this, so I don't think this is good or bad
2. We might end up going back to a capability-based approach if we need to scope the applicability of the time conductor. We've heard vague uses for wanting to scope this but don't have anything concrete for how this would look.

Based on (2) I slightly prefer the capability-based approach, but I'm fine with taking the simpler service-based approach, since those use cases aren't clear enough to say that capability decoration would really facilitate them anyway.

File: platform/features/conductor/src/TimeConductor.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+         * @memberof platform/features/conductor
+         * @constructor
+         * @augments {platform/telemetry.TelemetryCapability}
+         * @param {platform/features/conductor.TimeConductor} timeConductor
+         *        the time conductor which controls these queries
+         * @param {platform/telemetry.TelemetryCapability} telemetryCapability
+         *        the wrapped capability
+         */
+        function TimeConductor(start, end) {
+            this.inner = [ start, end ];
```
Comment (larkin) [inline platform/features/conductor/src/TimeConductor.js]: Would named properties (e.g. innerStart, innerEnd) convey more meaning here than array indices?

File: platform/features/conductor/src/ConductorRepresenter.js (hunk 2 above)
Comment (VWoeltjen) [inline platform/features/conductor/src/ConductorRepresenter.js]: I've hit https://docs.angularjs.org/error/ng/cpws on a few occasions when I've retained a reference via `this.scope`, including circumstances when I wasn't really expecting Angular to try to copy the object, so I've habituated revealing this via closure. That said, not sure this is necessary in a representer - I'll check

File: platform/commonUI/general/src/controllers/TimeConductorController.js [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+                $scope.startInnerPct =
+                    toPercent((ngModel.inner[0] - ngModel.outer[0]) / span);
+                $scope.endInnerPct =
+                    toPercent((ngModel.outer[1] - ngModel.inner[1]) / span);
+            }
+
+            function updateViewFromModel(ngModel) {
+                var t = now();
+
+                ngModel = ngModel || {};
```
Comment (larkin) [inline platform/commonUI/general/src/controllers/TimeConductorController.js]: Can we replace ngModel with an instance of TimeConductor and allow the TimeConductor constructor to set defaults?

like `ngModel = ngModel || new TimeConductor();`

File: platform/features/conductor/src/ConductorRepresenter.js (hunk 2 above)
Comment (VWoeltjen) [inline platform/features/conductor/src/ConductorRepresenter.js]: ...and after a quick check, it doesn't seem necessary in this case. Will change.

File: platform/features/conductor/src/ConductorTelemetryCapability.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            } else {
+                request.start = this.timeConductor.queryStart();
+                request.end = this.timeConductor.queryEnd();
+            }
+
+            return request;
+        };
+
+        ConductorTelemetryCapability.prototype.getMetadata = function () {
+            return this.wrappedCapability.getMetadata();
```
Comment (larkin) [inline platform/features/conductor/src/ConductorTelemetryCapability.js]: Hrm if we used inheritance we wouldn't have to write this-- would be a prototype lookup.  But if we do that we need to write a basic helper for inheritance (á la http://backbonejs.org/docs/backbone.html#section-247) but we don't have good story for how we'd reuse such a helper.  Will add that to my API notes.

File: platform/telemetry/src/TelemetrySubscription.js [hunk 6]
```diff
@@ -239,6 +239,10 @@ define(
 
             initialize();
             this.unlistenToMutation = addMutationListener();
+
+            // Expose makeDatum for TelemetryHandler to use, but not
+            // as part of public API
+            this.makeDatum = makeDatum;
```
Comment (larkin) [inline platform/telemetry/src/TelemetrySubscription.js]: Hrm, I don't think a method being on the prototype is necessarily the determinant of whether it's public or not-- would prefer to move all methods to prototype and tag them as `@private` for jsdoc (or adopt something like `_` prefix for `"private"`).

File: platform/features/conductor/src/ConductorTelemetryCapability.js (hunk 5 above)
Comment (VWoeltjen) [inline platform/features/conductor/src/ConductorTelemetryCapability.js]: One distinction here is that we're extending an _instance_, as opposed to a class. We could use `Object.create` and attach properties directly, or use closure scope to create a constructor whose prototype extends the instance, e.g.:

```js
function createTelemetryDecorator(telemetryCapability) {
    function ConductorTelemetryCapability() {}
    ConductorTelemetryCapability.prototype = Object.create(telemetryCapability);
    ConductorTelemetryCapability.prototype.requestTelemetry = ...
    return ConductorTelemetryCapability;
}
```

I prefer explicitly delegating in this case, for a few reasons:

* It's consistent with other code style (doesn't add helpers or deviate from the normal "here's a constructor with a prototype" approach)
* It's clear and explicit.
* Since the instance is delegated-to instead of prototypally-extended-from, there is not a risk of unintentionally overriding any methods used internally by the wrapped capability

File: platform/features/conductor/src/ConductorCapabilityDecorator.js (hunk 1 above)
Comment (larkin) [inline platform/features/conductor/src/ConductorCapabilityDecorator.js]: I think simpler is better given the murky-ness of non-global time conductor applicability.  

And to me, overriding capabilities adds an extra layer of code that becomes hard to follow.  Would rather make those overrides at the service layer -- with composite services -- instead of encouraging developers to override core platform functionality.

File: platform/features/conductor/src/ConductorTelemetryCapability.js (hunk 5 above)
Comment (larkin) [inline platform/features/conductor/src/ConductorTelemetryCapability.js]: Sure, due to the way composite services are assembled by the framework, this is extending an instance-- but it would be theoretically possible to have the framework use inheritance to assemble the service and instantiate only one instance of it.

The concern I have is the development cost of explicit delegation:
* more testing surface and more maintenance surface: if you forget to delegate required functionality, you break the entire service.  
* To get test coverage, you've got to test your delegation code.  
* If we were to add any methods to the service, every single decorator would have to be updated to also delegate that method.  

Inheritance would lower the amount of code and tests we have to write/maintain by utilizing language features.  As far as potentially overriding an internal method, well, that's a risk to watch out for... if we standardized internal method naming we could detect these conflicts at runtime.

There may be performance differences if there are any V8 optimizations that may exist for prototype-chain-lookups versus delegation by function calls, but that's not my primary concern.

Again-- a concern for framework/API design, not something to fix in this pull request.

File: platform/features/conductor/src/TimeConductor.js [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+ * queries and for time-based displays.
+ *
+ * @namespace platform/features/conductor
+ */
+define(
+    function () {
+        'use strict';
+
+        /**
+         * Wrapper for the `telemetry` capability which adds start/end
```
Comment (larkin) [inline platform/features/conductor/src/TimeConductor.js]: Looking at this JSDoc I'm starting to get a little confused by what this object is used for.  I'm guessing there's some cleanup here?

File: platform/features/conductor/src/TimeConductor.js (hunk 3 above)
Comment (VWoeltjen) [inline platform/features/conductor/src/TimeConductor.js]: Yep. Going with `inner.start` and `inner.end` etc. because it's a little easier to add some helpers around (particularly in the related controller & representer)

File: platform/commonUI/general/src/controllers/TimeConductorController.js (hunk 4 above)
Comment (VWoeltjen) [inline platform/commonUI/general/src/controllers/TimeConductorController.js]: Rationale for not doing this is to maintain a loose coupling between this visual control (which this controller supports) and time conductor as a feature. Corollary to this, will rename to `TimeRangeController`

Note that there's some additional logic for determining defaults in the `ConductorService`; these are the defaults that actually get used at run-time. Having this controller additionally enforce defaults is just a little bit of robustness to facilitate future reuse

**Comment by larkin** [comment]
I'm a bit concerned about how many changes this requires for represented views, ideally it's very loosely coupled to views, so that there is minimal work required for new views to integrate with the conductor.  I'm thinking whatever handler is called when $scope.domainObject changes could also be called whenever the time conductor changes.

That's it for a precursory review from me!  Thanks for getting this up early; it's easier to comprehend when I can start reviewing early.

File: platform/telemetry/src/TelemetrySubscription.js (hunk 6 above)
Comment (VWoeltjen) [inline platform/telemetry/src/TelemetrySubscription.js]: Agreed, and think we should adopt the `@private` approach generally (since adding JSDoc to private methods will facilitate static time-checking as well, and provides explicit disambiguation of public and private API).

File: platform/features/conductor/src/TimeConductor.js (hunk 7 above)
Comment (VWoeltjen) [inline platform/features/conductor/src/TimeConductor.js]: Regarding the class-level JSDoc, you are correct, that's a copy-paste failure. I will see if I can bind Ctrl-V to provide a mild electric shock, to encourage alertness.

Regarding the namespace-level JSDoc, this provides a bundle description; there's currently no clear place to put these since JSDoc has no concept of `bundle.json`, which I think is another point in favor of doing [bundle declarations in JavaScript](https://github.com/nasa/openmctweb/blob/api-redesign/docs/src/design/proposals/APIRedesign.md#bundle-declarations-in-javascript) (in which case we wouldn't need to repurpose `@namespace` as we have)

**Comment by VWoeltjen** [comment]
### Author Checklist

1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y <sup>1</sup>
3. Command line build passes? Y
4. Expect to pass code review? Y

<sup>1</sup> Unit tests are included but don't meet coverage requirements; would like to defer this with #116 to expedite integration of these changes

**Comment by VWoeltjen** [comment]
Note that platform/features/conductor is _not_ included in the set of active bundles; this is to keep it from appearing in the UI until after markup/styling tweaks, #117

**Comment by charlesh88** [comment]
Some questions as I dig into the styling in #117:
- I thought when we first looked at this that you were dynamically increasing or decreasing the number of horizontal time ticks as the user resizes the width of the view area. I don't see that behavior now... The reason I ask is that if we aren't dynamically doing this, then I have to treat how the tick label truncates differently.
- As I drag the time controller ends back and forth, the datetime label is really lagging behind. It really should be part of the element that is dragged visually, even if its value takes a bit of time to update. It's looking like the position of the datetime element isn't updating until the display loads, but that might just be an illusion. 

Assigning to Victor for responses.

**Comment by VWoeltjen** [comment]
@charlesh88 - That's strange, I'm not able to reproduce either of those behaviors locally. Are we looking at something different (different browser, different commit)? I'm viewing latest of open1515 in Chrome

**Comment by VWoeltjen** [comment]
Ah, I see both of those after I merge in latest from master. Guessing something from either theming or mobile is the problem? I'll investigate.

**Comment by charlesh88** [comment]
Belay the second "lagging" issue: as I refined the CSS for the look of the range bar, the problem went away. I still see the first problem though.

**Comment by VWoeltjen** [comment]
Okay, merged in latest from master.

1. The tick marks not updating was actually the `mct-resize` directive having broken. Not sure when this happened, possibly Angular upgrade? I've pushed a fix onto this branch.
2. The datetime label lag is, I believe, due to CSS animation. You can see the sliders animate into place when changed by using the date pickers (e.g. choose an End date a week into the future and watch the right hand slider animate over.) This effect looks great with the date pickers, but when you're using a knob it just looks like it's lagging behind. Is there a CSS solution for this (can the animation be disabled while the mouse is over the element that contains the knob/label?)

**Comment by charlesh88** [comment]
If you like, checkout open117 (in progress) which I just pushed.

**Comment by VWoeltjen** [comment]
@charlesh88 If you merge the latest from this branch, you should see the tick-update issue resolved.

**Comment by charlesh88** [comment]
1. I'll get latest from open1515 to get the fixed mct-resize. 
2. CSS animation as the problem totally makes sense here. My CSS mods to the slider probably removed a class that was supplying the animation directive. Drag works great now, but knobs and labels don't animate anymore. I'll look into how to bring it back when the user selects another datetime from the picker. We may have to do something like dynamically add a class to the element while the user has mousedown and is dragging that disables the animation.

**Comment by VWoeltjen** [comment]
One thing I'll note is that during development, I often found myself a little confused (and thinking there might be a bug) by the way the knobs would jump into place after using the date picker. I feel that the animation does help clarify what is going on, which I'd guess was the intent. Don't know that we necessarily need it, just noting that there's some observable utility in addition to the aesthetic value.

**Comment by larkin** [comment]
@charlesh88 have you tried using the `:active` pseudoselector for disabling the animation?  It typically applies whenever the primary mouse button is held down.

**Comment by charlesh88** [comment]
I have not, and that looks like a good idea. I'll check it out, thanks!

**Comment by VWoeltjen** [comment]
@larkin not sure if you caught the Author Checklist above - this is ready for review

**Comment by larkin** [comment]
Got it, will wrap up all Ymir reviews today.

Pete

On Sep 25, 2015, at 3:14 PM, Victor Woeltjen <notifications@github.com<mailto:notifications@github.com>> wrote:


@larkin<https://github.com/larkin> not sure if you caught the Author Checklist above - this is ready for review

File: platform/commonUI/general/src/directives/MCTPopup.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+ *****************************************************************************/
+/*global define*/
+
+define(
+    function () {
+        'use strict';
+
+        var TEMPLATE = "<div></div>";
+
+        function MCTPopup($window, $document, $compile) {
```
Comment (larkin) [inline platform/commonUI/general/src/directives/MCTPopup.js]: Any docstring for this component?  Seems to have functionality very similar to InfoService.

File: platform/features/conductor/src/ConductorRepresenter.js [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+        // Combine start/end times into a single object
+        function bounds(start, end) {
+            return { start: start, end: end };
+        }
+
+        // Update the time conductor from the scope
+        function wireScope(conductor, conductorScope, repScope) {
+            function updateConductorOuter() {
+                conductor.queryStart(conductorScope.conductor.outer.start);
```
Comment (larkin) [inline platform/features/conductor/src/ConductorRepresenter.js]: queryStart and queryEnd aren't used in the telemetry decorator or really any of the components, yet there is a lot of code in place to support them-- down to the level of watching conductor.outer.  Would be worth the large reduction in code and complexity to remove the unneeded code.

File: platform/features/conductor/src/ConductorTelemetrySeries.js [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+         * Bound a series of telemetry such that it only includes
+         * points from within the time conductor's displayable window.
+         *
+         * @param {TelemetrySeries} series the telemetry series
+         * @param {platform/features/conductor.TimeConductor} the
+         *        time conductor instance which bounds this series
+         * @constructor
+         * @implements {TelemetrySeries}
+         */
+        function ConductorTelemetrySeries(series, conductor) {
```
Comment (larkin) [inline platform/features/conductor/src/ConductorTelemetrySeries.js]: This entire wrapper seems unnecessary-- the underlying telemetry providers should not be providing data that is outside of the requested `start` and `end` dates.  If they do that, then the binary search that occurs here is worst case performance, every time.  Also forces all domainValues to be a date.

Also has the side effect of forcing the time conductor functionality onto all components-- no matter what the start date and end date were in the request, the time conductor will force the responses to exclude data outside that range.

File: platform/features/plot/src/PlotController.js [hunk 11]
```diff
@@ -165,6 +184,19 @@ define(
...
+            // Respond to a display bounds change (requery for data)
+            function changeDisplayBounds(event, bounds) {
+                self.pending = true;
+                releaseSubscription();
+                throttledRequery();
+                setBasePanZoom(bounds);
+            }
+
+            // Reestablish/reissue request for telemetry
+            throttledRequery = throttle(function () {
```
Comment (larkin) [inline platform/features/plot/src/PlotController.js]: Handling this throttling in one place (on the representer, before throwing the event) would result in not having to implement it in each view.  I'd even push for the time controller not to broadcast the event until the end of the user interaction.

Side note: How do you plan to handle request cancellation?  Server requests will queue up and will be served in the order they were queued; canceling those requests when they are no longer needed is key to ensuring performant view updates.

**Comment by larkin** [comment]
@VWoeltjen General telemetry api note, the multiple layers of functionality on top of the telemetry capability (telemetry handles/handlers/subscriptions/etc) is quite complex and unclear.

Is there a performance difference between directly using the telemetry capability and using the telemetry handles/subscriptions?

**Comment by larkin** [comment]
@VWoeltjen throwing back your way-- I think it's possible to strip a few hundred lines of code out of this pull request by removing code paths that are not used or are overstepping contractual boundaries.

File: platform/commonUI/general/src/directives/MCTPopup.js (hunk 8 above)
Comment (VWoeltjen) [inline platform/commonUI/general/src/directives/MCTPopup.js]: Good point, there's also very similar logic for context menus. I'll separate that out into a single place.

File: platform/features/plot/src/PlotController.js (hunk 11 above)
Comment (VWoeltjen) [inline platform/features/plot/src/PlotController.js]: Yes, throttling the event broadcasting is a much better approach to this. Waiting for the end of the user interaction still seems like overkill to me, but I think it will be fairly simple to wait for the bounds to become "stable" (e.g. user holds the bounds still for some number of ms) to get the desired effect. Think this will also remove the need for "request cancellation" since requests won't be issued until the bounds have stabilized.

If there are other cases where we specifically want to cancel out pending telemetry requests more generally, let's look at that in the context of #69.

File: platform/features/conductor/src/ConductorTelemetrySeries.js (hunk 10 above)
Comment (VWoeltjen) [inline platform/features/conductor/src/ConductorTelemetrySeries.js]: My thinking here was to avoid integration pain by reducing the extent to which providers needed to support these new request properties. I think this is particularly true for subscriptions, where filtering out data that falls out of the start/end times is both non-obvious, and something that shouldn't need to be done uniquely by each provider.

You're right that this does impose some additional constraints on domains; I'll take a second look and see if things look simpler without this.

**Comment by VWoeltjen** [comment]
Todos:

- [ ] Remove queryStart/queryEnd and related outer bounds tracking
- [x] Refactor out commonalities between mct-popup, info service, context menu
- [ ] Remove `ConductorTelemetrySeries` (or determine that it's really necessary)
- [ ] Move throttling into `ConductorRepresenter` (out of Plot view and/or other views)

**Comment by VWoeltjen** [comment]
@larkin - regarding layers for dealing with telemetry, yep, agreed. The utility services for dealing with telemetry are somewhat useful as facades, but would rather make it simpler to use the telemetry API at the capability level (or similar.) Think that discussion is better in the context of #69; there are a couple of proposed changes in there related to this, and I'd be quite interested in considering more

**Comment by VWoeltjen** [comment]
Pushed changes to address issues from code review, reassigning to @larkin for re-review.

**Comment by larkin** [comment]
# Review Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
5. Project-specific information isolated to appropriate branches? Y

# Integration Checklist

1. Automated build passing after merge? Y
2. No merge conflicts (or resolution trivial/obvious)? Y
3. All master branches up-to-date after merge? Y
4. Project-specific information isolated to appropriate branches? Y
